Added filter for minecraft server

MinePro120 opened the pull request:
Before submitting your PR, please review the following checklist:

- [x] **CHOOSE CORRECT BRANCH**: if filing a bugfix/enhancement
      against certain release version, choose `0.9`, `0.10` or `0.11` branch,
      for dev-edition use `master` branch
- [x] **CONSIDER adding a unit test** if your PR resolves an issue
- [x] **LIST ISSUES** this PR resolves
- [x] **MAKE SURE** this PR doesn't break existing tests
- [x] **KEEP PR small** so it could be easily reviewed.
- [x] **AVOID** making unnecessary stylistic changes in unrelated code
- [x] **ACCOMPANY** each new `failregex` for filter `X` with sample log lines
      within `fail2ban/tests/files/logs/X` file

MinePro120 committed:
Added [minecraft] to jail.conf

MinePro120 committed:
Added minecraft filter

MinePro120 committed:
Added newline at the end of minecraft.conf

MinePro120 commented:
The timestamp of the logs are in this format : "[21:20:18]", so I'm not sure what value would "time" in failJSON have.

sebres commented:
The time simulating by test suite for NOW is `1124013600` - `Sun, 14 Aug 12:00:00 CEST 2005` (in the time zone CET).
But because `[21:20:18]` lies in the future, fail2ban would assume failure from previous day, so it results to `2005-08-13T21:20:18`.

MinePro120 committed:
Added minecraft test log

sebres commented on config/jail.conf:
```diff
@@ -908,6 +908,14 @@ bantime      = 1h
+# Vanilla server, 1.16.3
+# Change port and logpath as needed
+port = 25565
+findtime = 20
+maxretry = 5
+bantime = 10d
```
All this values are rather a matter of user administrating fail2ban, so they are belonging to local settings, e. g. `jail.local`.
Please remove it.

MinePro120 replied:
Sure thing.

MinePro120 commented on config/filter.d/minecraft.conf:
```diff
@@ -0,0 +1,7 @@
+
+[Definition]
+
+failregex = \[\/<HOST>\:
```
There is no "access denied" really. There is a whitelist feature provided by the server, but it's rarely used. The purpose of this filter is to block a DOS attack in which players with random names join the server (an attack I've seen). So the key here are the findtime and maxretry options that can rule out spamming bots.

sebres replied:
`logged in` and `lost connection: Disconnected` surely don't represent any attack vector

bemk replied:
Here's what a bot attack looks like on my own server:
```
[03:32:54] [Server thread/INFO]: com.mojang.authlib.GameProfile@508346c3[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:50306) lost connection: Disconnected
[03:36:28] [Server thread/INFO]: com.mojang.authlib.GameProfile@4392677b[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:54278) lost connection: Disconnected
[03:39:37] [Server thread/INFO]: com.mojang.authlib.GameProfile@59032ddf[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:59320) lost connection: Disconnected
[03:40:53] [Server thread/INFO]: com.mojang.authlib.GameProfile@3c65e0e1[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:55888) lost connection: Disconnected
[03:46:07] [Server thread/INFO]: com.mojang.authlib.GameProfile@672c7a59[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:46164) lost connection: Disconnected
[03:56:35] [Server thread/INFO]: com.mojang.authlib.GameProfile@3feaeec1[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:47884) lost connection: Disconnected
[04:00:12] [Server thread/INFO]: com.mojang.authlib.GameProfile@124d287d[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:47432) lost connection: Disconnected
[04:02:28] [Server thread/INFO]: com.mojang.authlib.GameProfile@4c7d9ef5[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:38110) lost connection: Disconnected
[04:06:09] [Server thread/INFO]: com.mojang.authlib.GameProfile@72fd24a0[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:47528) lost connection: Disconnected
[04:09:48] [Server thread/INFO]: com.mojang.authlib.GameProfile@62e80648[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:59492) lost connection: Disconnected
[04:13:36] [Server thread/INFO]: com.mojang.authlib.GameProfile@5981520a[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:35350) lost connection: Disconnected
[04:17:27] [Server thread/INFO]: com.mojang.authlib.GameProfile@351d996e[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:49122) lost connection: Disconnected
[04:21:19] [Server thread/INFO]: com.mojang.authlib.GameProfile@62be7d0d[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:50928) lost connection: Disconnected
[04:25:02] [Server thread/INFO]: com.mojang.authlib.GameProfile@e1ada86[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:59308) lost connection: Disconnected
[04:28:42] [Server thread/INFO]: com.mojang.authlib.GameProfile@46d2bd38[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:44334) lost connection: Disconnected
[04:32:41] [Server thread/INFO]: com.mojang.authlib.GameProfile@4cf80fb3[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:58452) lost connection: Disconnected
[04:36:37] [Server thread/INFO]: com.mojang.authlib.GameProfile@34b3d11a[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:58510) lost connection: Disconnected
[04:40:34] [Server thread/INFO]: com.mojang.authlib.GameProfile@29a1059[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:37508) lost connection: Disconnected
[04:44:27] [Server thread/INFO]: com.mojang.authlib.GameProfile@240239f7[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:56702) lost connection: Disconnected
[04:48:26] [Server thread/INFO]: com.mojang.authlib.GameProfile@4d26720b[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:55874) lost connection: Disconnected
[04:52:23] [Server thread/INFO]: com.mojang.authlib.GameProfile@6e4a5cf3[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:34900) lost connection: Disconnected
[04:56:20] [Server thread/INFO]: com.mojang.authlib.GameProfile@1457bf54[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:32780) lost connection: Disconnected
[05:00:11] [Server thread/INFO]: com.mojang.authlib.GameProfile@213952cc[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:37350) lost connection: Disconnected
[05:04:07] [Server thread/INFO]: com.mojang.authlib.GameProfile@342c9e32[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:40264) lost connection: Disconnected
[05:08:00] [Server thread/INFO]: com.mojang.authlib.GameProfile@4c6b588a[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:42654) lost connection: Disconnected
[05:11:57] [Server thread/INFO]: com.mojang.authlib.GameProfile@40ec9934[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:42936) lost connection: Disconnected
[05:16:17] [Server thread/INFO]: com.mojang.authlib.GameProfile@19967be1[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:54126) lost connection: Disconnected
[05:20:11] [Server thread/INFO]: com.mojang.authlib.GameProfile@151407c8[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:44118) lost connection: Disconnected
[05:24:03] [Server thread/INFO]: com.mojang.authlib.GameProfile@1b7635c7[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:60018) lost connection: Disconnected
[05:28:02] [Server thread/INFO]: com.mojang.authlib.GameProfile@5fcdd47a[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:60892) lost connection: Disconnected
[05:31:57] [Server thread/INFO]: com.mojang.authlib.GameProfile@6c78fdce[id=<null>,name=cuute,properties={},legacy=false] (/20.4.48.76:60474) lost connection: Disconnected

```

It's a whitelisted server using online mode (meaning player identities are verified by Mojang). As a result the attackers are only generating a disconnected message without ever generating the messages:

```
[14:36:24] [Server thread/INFO]: <player name>[/<IP>:<port>] logged in with entity id 73 at (-611.530629608662, 65.0, -266.9221458299592)
[14:36:24] [Server thread/INFO]: <player name> joined the game
```

Regular disconnect messages come in a couple of forms:

```
[14:43:01] [Server thread/INFO]: <player name> lost connection: Disconnected
[14:43:01] [Server thread/INFO]: <player name> left the game
```
```
[19:54:35] [Server thread/INFO]: <player name>lost connection: Timed out
[19:54:35] [Server thread/INFO]: <player name> left the game
```

This last disconnect, noted here below, came seconds before the legitimate player decided to log in, suggesting this is some kind of scan by the legitimate client to gather server status information on the list of servers configured by the player. This information contains server connection quality as well as the list of players currently online.

The top line of this one makes it difficult to just take the contents of this specific line to make a filter, given the fact that it's an exact match for the log result when a bot scans your server.
```
[11:21:19] [Server thread/INFO]: com.mojang.authlib.GameProfile@27ce9407[id=<null>,name=<player name>,properties={},legacy=false] (/<ip>:<port>) lost connection: Disconnected
[11:21:29] [User Authenticator #2/INFO]: UUID of player <player name> is <player uuid>
[11:21:29] [Server thread/INFO]: <player name>[/<ip>:<port>] logged in with entity id 294 at (-45.820534318481705, 68.0, 60.00821373182576)
[11:21:29] [Server thread/INFO]: <player name> joined the game
```

The main purpose of the attack seems to be people scanning for player online information to find griefing targets, or to find servers with sensitive player names (such as youtubers) so they can attack those servers.

trainzkid replied:
What's the difference between a bot's `lost connection: Disconnected` message in the logs and a legitimate player's? If they're the same, or have any crossover, fail2ban can't protect against that afaik.

bemk replied:
@trainzkid whilst it is possible for a legitimate user to generate the messages containing the `com.mojang.authlib.GameProfile` bits, they usually have a much simpler disconnect message. So it's really the frequency of the authlib line that stands out.

If you look at my attack log, at the 4 AM mark, you see that an abnormal connection disconnect happens like 16 times within that hour alone. If that were a normal player trying to put some extra load on my server by trying to load and unload the world around them that often I'd also have no issue with fail2ban showing them the door for a couple of hours/days.

As for when normal players create that message? I've seen it happen once, maybe twice a week. Could be a network issue when that happens though, and if that's the case you don't really wanna block them for having a bad ISP.

With all that said, I agree that Mojang could be providing better log messages for this kind of issue.

trainzkid replied:
I see what you mean, however, I think the main intent of fail2ban is to "ban" on bad authentication type errors, and the suggested filter above is just way too vague/broad that it's just as likely to affect legitimate players as it is bots.

That being said, the beauty of open source software and the freedom of config files means anyone who disagrees with my statement is more than welcome to copy the above filter and go to town on their local Minecrafters with extreme prejudice.

I also agree that Mojang could definitely improve their error messages a bit better so it's more obvious when a legitimate bot/baddie attempts to access your server. Unfortunately, I'm not confident that Mojang will make any sort of improvements in the near future. All we can do is adapt with what we're given.

On another note, adding a filter for the whitelist error doesn't sound like a bad idea to me. I think that probably has more candidacy than the subtle "disconnected" error, personally.

MinePro120 committed:
Removed [minecraft] findtime, maxretry

MinePro120 committed:
[minecraft] support for whitelist and auth

-Added support for blocking non-whitelisted players
-Added support for blocking "cracked" players
-Changed jail.conf recommendations

MinePro120 committed:
[minecraft] updated test log

Ludo-code commented:
Any news about this pull request?

sebres commented on fail2ban/tests/files/logs/minecraft:
```diff
@@ -0,0 +1,2 @@
+# failJSON: { "time": "2005-08-13T21:30:40", "match": true, "host": "123.45.67.89" }
+[21:30:40] [Server thread/INFO]: Player[/123.45.67.89:12345] logged in with entity id 551 at (825.302579326729, 65.0, -886.9107800126116)
```
So you decided to remove it from the test, but the issue remain - the message will be matched with the filter, so still don't think such regex is good idea.

tyl0re replied:
For Geyser (de-DE) these Message should also be handled:
[epollEventLoopGroup-3-2/INFO]: [Geyser-Spigot] /123.127.235.6:1051 versuchte zu verbinden

sebres commented:
I don't think it is justified to merge this PR, because it still looks "vulnerable" at the moment.
Moreover, take a closer look to testcase - it looks like a single attempt will be counted 3 times (as it'd be 3 attempts), what is not correct in my opinion.

If the log-file containing the message like this: https://github.com/fail2ban/fail2ban/blob/778b09404f2885252088f6feb9ec3854508615bc/fail2ban/tests/files/logs/minecraft#L4
I would use something like this instead:
```ini
[minecraft]
port = ...
logpath = ...
filter = 
failregex = ^[^:]*: Disconnecting \S+ \(/<ADDR>:\d+\): You are not white-listed on this server!
enabled = true
```

Omig12 commented:
This is still a non-working solution, but it builds on @sebres suggestion:

```sh 
failregex = \[(?P<time>\d+:\d+:\d+)\].*\(\/(?P<host>\d+.\d+.\d+.\d+):\d+\) lost connection.*$
```

The host and time capture attempt to comply with some [fail2ban specifications](https://www.fail2ban.org/wiki/index.php/MANUAL_0_8#:~:text=for%20%22permanent%22%20ban.-,Filters,-The%20directory%20filter), but Minecraft Servers output the timestamp in a format that [seems incompatible with the fail2ban time regex](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=491253%7C). Which I think is what @MinePro120 and @sebres were discussing earlier. I'm looking for a way to modify how the Minecraft server reports the date and time in the logs. Still, all I've found so far are [guides for tweaking log4j](https://www.reddit.com/r/admincraft/comments/69271l/guide_controlling_console_and_log_output_with/); from my understanding, that logger was deprecated for vulnerabilities.

sebres commented:
> This is still a non-working solution

Why then do you provide such non-working "solution" here?

> I'm looking for a way to modify how the Minecraft server reports the date and time in the logs.

You don't need... One can simply specify [datepattern](https://manpages.debian.org/bullseye/fail2ban/jail.conf.5.en.html#datepattern) for the jail or filter.
It looks like you'd need:
```ini
datepattern = ^\[%%H:%%M:%%S\]\s+
```

By the way - usage of catch-all's is ugly and `.*$` is not an anchor at all, so either remove it completely (if it is justified unanchored) or rewrite catch-all more precisely.
And don't use captured groups of `(?P<host>...)`, use `<ADDR>` (or `<HOST>`) instead.

Omig12 commented:
Hey, thanks for the reply. 

> Why then do you provide such non-working "solution" here?

Because this is a collaborative platform, and instead of being snarky, I'll contribute my limited knowledge so that others may improve upon it.

> You don't need... One can simply specify [datepattern](https://manpages.debian.org/bullseye/fail2ban/jail.conf.5.en.html#datepattern) for the jail or filter.

The date patterns the logs provide resolve from seconds up to the hour, so there is no day, month, or year information. Thanks for that link to the documentation; I see that by specifying datepattern as NONE we could bypass that limitation. 

I don't know what the fial2ban best practices are. Would you recommend keeping the datepattern matching the time from seconds up to the hour or setting it to NONE? 

Regarding the use of catch-alls not sure how to avoid using it since the log entries are quite heterogeneous. It might not be the best solution, but if it works, it provides something for others with the skills to improve upon it. I was testing it against the regex 101 Python option, and `.*$` seemed to work; I might not be understanding your comment correctly.

> And don't use captured groups of (?P<host>...), use <ADDR> (or <HOST>) instead.

Just followed what the DOCs I had available said.

![image](https://github.com/fail2ban/fail2ban/assets/6131141/2e970f4a-6ac6-4e1f-99aa-2d8e04314e1b)

https://www.fail2ban.org/wiki/index.php/MANUAL_0_8#:~:text=for%20%22permanent%22%20ban.-,Filters,-The%20directory%20filter

Hope this helps anyone trying to improve on this solution, it would be really cool to have something like this working (even if it's just remedially) to get rid of the robot-spam on the server.

trainzkid commented:
This isn't _quite_ the same as the use case mentioned in the aforementioned PR, but I've found using a systemd service for logging means fail2ban can utilize systemd-journal's timestamps, and a regex that matches the _whitelist_ errors works. That's not exactly the bot problem this discussion has been about, but it's better than nothing. 

Here's my `minecraft-whitelist.conf`:
```
[INCLUDES]
before = common.conf

[Definition]
failregex = Disconnecting .* \(/<ADDR>:.*\): You are not white-listed on this server!
journalmatch = _SYSTEMD_UNIT=$SERVICE.service + _COMM=$START.sh
```
where `$SERVICE` is the name of the systemd service and `$START.sh` is the script or binary that starts the server (usually `ServerStart.sh` or something similar).

I'm also aware there aren't any anchors, I've had some trouble with them in the past for whatever reason.

sebres commented:
> I'm also aware there aren't any anchors, I've had some trouble with them in the past for whatever reason.

Well it depends on log-format (and log-target, e. g. systemd journal or log-file), here is fully-anchored example for log-line from this PR:
```ini
failregex = ^[^:]*: Disconnecting [^\(]*\(/<ADDR>:\d+\): You are not white-listed on this server!
```
(`[^:]*` and `[^\(]*` are not catch-alls and mean everything before colon and open round parenthesis, respectively)

Alternatively it can be only anchored from the end, with few conditions:
- there should be no catch-alls between `<ADDR>` and determining message (exactly the case here);
- logging of free message content (some foreign user input) at the end of the log is explicitly excluded (hardly possible to estimate without to review the source code or asking developers).

```ini
failregex = \(/<ADDR>:\d+\): You are not white-listed on this server!\s*$
```

trainzkid commented:
I think the issues I ran into were related to extraneous timestamps at the beginning of a log entry, despite using systemd-journal. Basically, the timestamp mentioned earlier in this thread is still present in the journal entry, so when I tried to anchor at the beginning, I'm guessing my regex was having trouble with that initial timestamp when it had already successfully parsed the journal's timestamp. I don't know if fail2ban tries to parse timestamps again after successfully getting the timestamp from the journal or if I need to add something to the regex to make it ignore that initial section in square brackets with the timestamp. I can't tell if that's what your regex does, it doesn't look like it does. 

The other issue I may have been facing was when I heard anchoring, I thought it meant on _both_ sides (both beginning _and_ end). Is it preferred to only anchor from one or the other and not both? 

I think in this specific case for `minecraft-whitelist`, anchoring from the end would work just fine, I don't think there is any added input in between.

sebres commented:
>  I thought it meant on both sides (both beginning and end). 

It is true, either if one needs to be precise as possible, or if there are some catch-alls similar REs used inside, before the captured data in left anchored RE, or after captured data in right anchored RE (quasi to restrict corner cases for injection attempts via foreign data).

> Is it preferred to only anchor from one or the other and not both?

It depends, but generally spoken - no, it is not preferred. 
If there is no catch-alls and matched message part is obvious a failure and such match allows unambiguously distinguish failed messages from the rest, one can surely use the anchor from one side only (mostly begin-anchor).
This has also another pros, like it wouldn't matter for which reason the failure happens, it is still a failure and it'd be considered as a failed attempt regardless what devs of service would devise later (or which info the log may get latter additionally).

For instance an RE `^\s*authentication failed for user "[^"]+":` would match all that:
```
authentication failed for user "usr1": user unknown
authentication failed for user "usr2": password mismatch
authentication failed for user "usr3": auth-service unavailable
authentication failed for user "usr4": unsupported method
authentication failed for user "usr5": whatever reason devs invent tomorrow or more data gets included later
```
And if something of that shall not cause a failure (e. g. 3rd message), it can be pretty simple done with negative lookahead: `^\s*authentication failed for user "[^"]+": (?!auth-service unavailable)`

It is hardly possible to give an unique answer for general usecase.

> I think in this specific case for minecraft-whitelist, anchoring from the end would work just fine, I don't think there is any added input in between.

"I don't think" is not equal "I know", what is mostly expected for a stock filter.
You can basically do what you want on your hosts (since it is your responsibility). 
But in case of some OSS-project, we must ensure it'd not cause false positives (or be affected by some inject-attempts).

ComputerCraftr commented:
This might not be the cleanest but it works with grep -E to find spam log entries on an english vanilla 1.21.4 server:
`^\[([01][0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9]\] \[Server thread\/INFO\]: [a-zA-Z0-9_]{3,16}( \(|\[)\/((25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9])\.){3}(25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9]):(6553[0-5]|655[0-2][0-9]|65[0-4][0-9]{2}|6[0-4][0-9]{3}|[1-5]?[0-9]{1,4})(\)|\]) (logged in with entity id [0-9]+ at \(-?[0-9]*\.[0-9]*, -?[0-9]*\.[0-9]*, -?[0-9]*\.[0-9]*\)|lost connection: (Disconnected|You are not white-listed on this server!|You are banned from this server.))$`

Test cases on my server with the ServerSeeker bot:
```
[03:27:12] [Server thread/INFO]: ServerSeekerV2 (/46.101.233.169:40550) lost connection: Disconnected
[11:12:35] [Server thread/INFO]: ServerSeekerV2 (/46.101.233.169:57416) lost connection: Disconnected
[16:21:17] [Server thread/INFO]: John8686[/10.0.0.1:56002] logged in with entity id 191 at (3.914456420848696, 69.0, -105.95373721049872)
[15:16:44] [Server thread/INFO]: John8686[/10.0.0.1:50287] logged in with entity id 174 at (7.5, 69.0, -97.5)
[16:21:24] [Server thread/INFO]: John8686[/10.0.0.1:50841] logged in with entity id 364 at (8.048320255974, 69.0, -100.81653781685961)
[16:21:54] [Server thread/INFO]: John8686[/10.0.0.1:50848] logged in with entity id 563 at (8.28792924518966, 69.0, -101.54784785831826)
[16:53:37] [Server thread/INFO]: John8686[/10.0.0.1:51064] logged in with entity id 781 at (8.28792924518966, 69.0, -101.54784785831826)
[16:54:02] [Server thread/INFO]: John8686 (/10.0.0.1:51082) lost connection: You are not white-listed on this server!
[16:54:09] [Server thread/INFO]: John8686 (/10.0.0.1:51084) lost connection: You are not white-listed on this server!
[17:01:56] [Server thread/INFO]: John8686[/10.0.0.1:51108] logged in with entity id 1005 at (8.28792924518966, 69.0, -101.54784785831826)
[17:23:27] [Server thread/INFO]: John8686 (/10.0.0.1:51253) lost connection: You are banned from this server.
[17:25:03] [Server thread/INFO]: John8686[/10.0.0.1:51258] logged in with entity id 1246 at (7.514844286305745, 69.0, -101.53915589835108)
```

skycatminepokie commented:
I think this would work well as a Minecraft mod/plugin. Malicious chat messages that are sent in funny ways could get ips banned.

rharish101 commented:
> This might not be the cleanest but it's anchored and works with grep -E to find spam log entries on an english vanilla 1.21.4 server: `^\[([01][0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9]\] \[Server thread\/INFO\]: [a-zA-Z0-9_]{3,16}( \(|\[)\/((25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9])\.){3}(25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9]):(6553[0-5]|655[0-2][0-9]|65[0-4][0-9]{2}|6[0-4][0-9]{3}|[1-5]?[0-9]{1,4})(\)|\]) (logged in with entity id [0-9]+ at \(-?[0-9]*\.[0-9]*, -?[0-9]*\.[0-9]*, -?[0-9]*\.[0-9]*\)|lost connection: (Disconnected|You are not white-listed on this server!|You are banned from this server.))$`
> 
> Test cases on my server with the ServerSeeker bot:
> 
> ```
> [03:27:12] [Server thread/INFO]: ServerSeekerV2 (/46.101.233.169:40550) lost connection: Disconnected
> [11:12:35] [Server thread/INFO]: ServerSeekerV2 (/46.101.233.169:57416) lost connection: Disconnected
> [16:21:17] [Server thread/INFO]: John8686[/10.0.0.1:56002] logged in with entity id 191 at (3.914456420848696, 69.0, -105.95373721049872)
> [15:16:44] [Server thread/INFO]: John8686[/10.0.0.1:50287] logged in with entity id 174 at (7.5, 69.0, -97.5)
> [16:21:24] [Server thread/INFO]: John8686[/10.0.0.1:50841] logged in with entity id 364 at (8.048320255974, 69.0, -100.81653781685961)
> [16:21:54] [Server thread/INFO]: John8686[/10.0.0.1:50848] logged in with entity id 563 at (8.28792924518966, 69.0, -101.54784785831826)
> [16:53:37] [Server thread/INFO]: John8686[/10.0.0.1:51064] logged in with entity id 781 at (8.28792924518966, 69.0, -101.54784785831826)
> [16:54:02] [Server thread/INFO]: John8686 (/10.0.0.1:51082) lost connection: You are not white-listed on this server!
> [16:54:09] [Server thread/INFO]: John8686 (/10.0.0.1:51084) lost connection: You are not white-listed on this server!
> [17:01:56] [Server thread/INFO]: John8686[/10.0.0.1:51108] logged in with entity id 1005 at (8.28792924518966, 69.0, -101.54784785831826)
> [17:23:27] [Server thread/INFO]: John8686 (/10.0.0.1:51253) lost connection: You are banned from this server.
> [17:25:03] [Server thread/INFO]: John8686[/10.0.0.1:51258] logged in with entity id 1246 at (7.514844286305745, 69.0, -101.53915589835108)
> ```

There's another case: when a player in online mode with an invalid username wants to join. I tried to log in to my PaperMC server using an online mode account with an invalid username (I tried longer than 16 and special characters) through Prism launcher, and I got logs like this:
```
[19:34:21] [Server thread/INFO]: /212.203.51.45:62187 lost connection: Disconnected
[19:37:03] [Server thread/INFO]: /212.203.51.45:6681 lost connection: Disconnected
[19:37:27] [Server thread/INFO]: /212.203.51.45:62171 lost connection: Disconnected
```

Note the missing username.

ComputerCraftr commented:
> There's another case: when a player in online mode with an invalid username wants to join. I tried to log in to my PaperMC server using an online mode account with an invalid username (I tried longer than 16 and special characters) through Prism launcher, and I got logs like this:
> 
> ```
> [19:34:21] [Server thread/INFO]: /212.203.51.45:62187 lost connection: Disconnected
> [19:37:03] [Server thread/INFO]: /212.203.51.45:6681 lost connection: Disconnected
> [19:37:27] [Server thread/INFO]: /212.203.51.45:62171 lost connection: Disconnected
> ```
> 
> Note the missing username.

Good catch! This one should include those without parentheses or usernames:
`^\[([01][0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9]\] \[Server thread\/INFO\]: [a-zA-Z0-9_]{0,16}( \(|\[)?\/((25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9])\.){3}(25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9]):(6553[0-5]|655[0-2][0-9]|65[0-4][0-9]{2}|6[0-4][0-9]{3}|[1-5]?[0-9]{1,4})(\)|\])? (logged in with entity id [0-9]+ at \(-?[0-9]*\.[0-9]*, -?[0-9]*\.[0-9]*, -?[0-9]*\.[0-9]*\)|lost connection: (Disconnected|You are not white-listed on this server!|You are banned from this server.))$`
I could put this in a separate PR if it seems good enough to catch most cases and won't have false positives.

sebres commented:
Oh god, what a mess... one of much weird logging formats I ever saw.

> This might not be the cleanest but it's anchored and works with grep -E to find spam log entries on an english vanilla 1.21.4 server

Why would one need `grep -E` here?
```bash
# to get all matched messages:
fail2ban-regex -o msg /path/to/log minecraft
# to get all unique IPs:
fail2ban-regex -o ip /path/to/log minecraft | sort -u
# to get all unique IP:user pairs (for RE provided below):
fail2ban-regex -o '<ip> : <F-USER>' /path/to/log minecraft | sort -u
# to get all unique IP:failure-reason pairs (for RE provided below):
fail2ban-regex -o '<ip> : <F-REASON>' /path/to/log minecraft | sort -u
```

> This one should include those without parentheses or usernames ...

Hmm, why not something like this:
```ini
datepattern = ^\[%%H:%%M:%%S\]
failregex = ^\s*\[Server thread\/INFO\]: (?:<F-USER>[^\s/]+</F-USER> ?)?[\[\(]?/<ADDR>:\d{1,5}[\]\)]? <F-REASON>(?:logged in with entity id|lost connection: (?:Disconnected|You are not white-listed on this server|You are banned from this server))</F-REASON>\b
```
or better readable/structured (with better potential for localization):
```ini
datepattern = ^\[%%H:%%M:%%S\]
_reasons = (?:logged in with entity id|lost connection: (?:Disconnected|You are not white-listed on this server|You are banned from this server))
failregex = ^\s*\[Server thread\/INFO\]: (?:<F-USER>[^\s/]+</F-USER> ?)?[\[\(]?/<ADDR>:\d{1,5}[\]\)]? <F-REASON>%(%_reasons)s</F-REASON>\b
```

> it seems good enough to catch most cases and won't have false positives.

I'm not sure this or even my above are "good enough" (for PR)... For instance, nobody still answered my question in https://github.com/fail2ban/fail2ban/pull/2852#discussion_r505803133
The same is valid about "won't have false positives".
Although I'm not familiar with minecraft and may misunderstand its log output or any "attack" vector or the needs (e.g. stopping flood or whatever).
For personal use it may be indeed good enough, at least if one understands what one does and can estimate the risks.

ComputerCraftr commented:
> Why would one need `grep -E` here?

Ah whoops, I mostly work on macOS and FreeBSD systems, so it's a force of habit reading logs with BSD grep...
> or better readable/structured (with better potential for localization):
> 
> ```ini
> datepattern = ^\[%%H:%%M:%%S\]
> _reasons = (?:logged in with entity id|lost connection: (?:Disconnected|You are not white-listed on this server|You are banned from this server))
> failregex = ^\s*\[Server thread\/INFO\]: (?:<F-USER>[^\s/]+</F-USER> ?)?[\[\(]?/<ADDR>:\d{1,5}[\]\)]? <F-REASON>%(%_reasons)s</F-REASON>\b
> ```

Yes, it would be a good idea to break out all of the strings into variables while we keep the same structure for parsing the logs. Then it will be easier to customize for other languages, different server messages, etc
> I'm not sure this or even my above are "good enough" (for PR)... For instance, nobody still answered my question in [#2852 (comment)](https://github.com/fail2ban/fail2ban/pull/2852#discussion_r505803133) The same is valid about "won't have false positives". Although I'm not familiar with minecraft and may misunderstand its log output or any "attack" vector or the needs (e.g. stopping flood or whatever). For personal use it may be indeed good enough, at least if one understands what one does and can estimate the risks.

With regards to the player logged in message, it would probably be good to block users who are joining/leaving quickly to spam join/leave messages in chat, and I did see it from some scanner traffic before as well:
```
[03:54:45] [Server thread/INFO]: kons0las (/199.58.83.10:53024) lost connection: Disconnected
[09:02:30] [User Authenticator #1/INFO]: UUID of player kons0las is bc4f5a10-e49d-4b95-b50e-5e5b108166dd
[09:02:31] [Server thread/INFO]: kons0las[/199.58.83.10:44776] logged in with entity id 155 at (-42.5, 74.0, -114.5)
[09:02:31] [Server thread/INFO]: kons0las joined the game
[09:02:32] [Server thread/INFO]: kons0las lost connection: Disconnected
[09:02:32] [Server thread/INFO]: kons0las left the game
```
Some bots have put discord invite links in their username and spammed the chat with join/leave messages for advertising, that is also something I noticed since the discord invite tokens are all alphanumeric and valid usernames.

However, this regex shouldn't have false positives - actual messages typed in the game chat have usernames wrapped in < > angle brackets and the server announcement uses [ ] square brackets, neither of which are valid characters in a Minecraft username, see my console output here:
```
[19:29:42] [Server thread/INFO]: John8686[/[2600:1011:b10b:678a:85f:5065:e126:f566]:64846] logged in with entity id 208 at (8.471545332800037, 66.0, -100.3912730775159)
[19:29:42] [Server thread/INFO]: John8686 joined the game
[19:30:01] [Server thread/INFO]: <John8686> Hello!
say Hello!
[19:30:06] [Server thread/INFO]: [Not Secure] [Server] Hello!
msg John8686 Hello!
[19:30:43] [Server thread/INFO]: You whisper to John8686: Hello!
[19:31:56] [Server thread/INFO]: John8686 lost connection: Disconnected
[19:31:56] [Server thread/INFO]: John8686 left the game
```
We will more than likely have false negatives, but really the false positives are probably what we would want to eliminate before merging any code and then we could just focus on basic vanilla and paper servers. If a server is already heavily modified, then the operators should probably install another plugin for application layer bot filtering or customize the fail regex to suit their needs instead.
